README
SPECTRUM 3D: 3D AUDIO SPECTRUM ANALYSER IN REAL TIME

Spectrum 3D dispays a 3D audio spectrogram in real time; the source can be either the microphone or an audio file or optionally jack (jack-audio-connection-kit). Optionally, it supports multitouch gestures from touchscreen and touchpad. Optionally also, the audio stream can be run in realtime even if Jack is not used. It is build with the Gstreamer, SDL, OpenGl, GTK+-2.0 and uTouch-Geis free libraries and is under GPL license. It is written for Ubuntu but should work for other Linux distributions except maybe for the multitouch abilities (the uTouch library use is not yet very common). 

If you want multitouch support, you should read the notes at the bottom of the page. A tutorial is provided in the 'Doc' directory.

*******************
|  INSTALLATION:  |
*******************

1. dependencies:
""""""""""""""""
FOR UBUNTU 10.10 and 11.04 : All dependencies are available in the Ubuntu repositories.

		- gcc, pkg-config, automake;
		- libgtk+-2.0-dev;
		- lidsdl1.2-dev and libsdl-ttf2.0-dev;
		- libgstreamer0.10-dev;

		- gstreamer0.10-plugins-bad, gstreamer0.10-plugins-bad-multiverse, gstreamer0.10-plugins-ugly et gstreamer0.10-plugins-ugly-multiverse (for reading files such as mp3);

	For JACK support : 
		- libjack-dev (for jack1) or libjack-jackd2-dev for jack2; note that you don't need to have the 'jackd' package installed, but only the '-dev' library;

	For multitouch support (touchscreen and also for multitouch touchpad from Natty):
		- libutouch-geis-dev version 2. For Ubuntu Oneiric Ocelot (11.10), use the packages from repositories. For Ubuntu Natty Narwhal (11.04), the repositories version will work, but filtering on number of touches won't work (Bug#769859 in libutouch-geis); to solve this you need a version of uTouch-geis that is at least 2.1.0; you can have it either by installing the '.deb' packages from Oneiric repositories (type in a terminal 'sudo apt-get build-dep libutouch-geis-dev', then install 'libutouch-geis1' first, then 'libutouch-geis-dev'), or by compiling the latest version from source (https://launchpad.net/utouch-geis); if you compile from sources, you can type in a terminal 'sudo apt-get build-dep libutouch-geis-dev' to have all dependecies installed. For Ubuntu Maverick (10.10), do not use the repositories version; use the solutions proposed for Natty. I don't know if multitouch support can easily be implemented in Ubuntu Lucid Lynx (10.04) or earlier.

FOR OTHER DISTRIBUTIONS : 
	For the other distributions, the dependencies should be very similar. Only the multitouch features maybe won't work since uTouch-Geis has been build for Ubuntu. However, the uTouch Suite (including uTouch-Geis) starts beeing ported in other distributions such as Gentoo (https://sabotageandi.wordpress.com/2011/04/20/utouch-gentoo/) or Fedora. More info on this in the documentation directory.

2. compilation and installation : 
"""""""""""""""""""""""""""""""""
 	1) in a terminal, go to the directory where the sources are; for example, if the source package is in the HOME directory :

		cd ~/spectrum3d-the-actual-version (if the sources are in the '/home/victor' directory);

	2) type : 

		./configure 

	If 'libjack-dev' or 'libjack-jackd2-dev' is found, JACK support will be enabled. Also, if 'libbamf-dev' and 'libutouch-geis-dev' version 2 are found, MULTITOUCH support will be enabled. If those libraries are not installed, JACK and MULTITOUCH support are not enabled. However, if you have those libraries installed but don't want those features enabled, you have to type '--disable-jack' or '--disable-geis' after 'configure' ('--enable-jack=no' and '--enable-geis=no' also works).
	For example :	

		./configure --disable-jack --disable-geis

will deactivate JACK and MULTITOUCH support if those libraries are installed.

	If you want Spectrum3d to be installed in another directory than the default one ('/usr/local/bin'), you can add it at 'configure' step according to the 'Autoconf' standards. For example : 

		./configure --prefix=/usr

will install the executable 'spectrum3d' file in '/usr/bin' directory instead of '/usr/local/bin'. 

	3) type :

		make

	4) type : 

		sudo make install (the password will be necessary for this)

	5) to completely remove Spectrum3d, type :

		sudo make uninstall


*********
|  USE  |
*********

Run it from a terminal by typing : 

	spectrum3d

or from the 'Menu->Applications->Sound & Video'

From the graphical interface, you can play/pause or stop playing. For audio files, you can seek forward or backward (5 seconds by 5 seconds); selection of multiple audio files is possible. 

You can play a test sound which is a sine wave from 1 to 20000 Hz.

Audio source can be microphone, audio file or jack (if you have enabled JACK support). You can change the speed of refreshing of the display. You can show a window that allows to control an equalizer and a band-pass (BP) or band-reject (BR) filter. For the BP/BR filter, you can set the lower and upper frequencies. For the eqalizer, you can adjust the gain (from -72dB to +36dB); for each band, you can change the frequency (from 20 to 20000 Hz) it works on and its the bandwidth (from 0 to 20000 Hz). 

The lowest displayed frequency is 0 Hz by default, and it can be increased. The default range is 2000 Hz. It can be decreased to 40 Hz or raised to 20000Hz. To change this, you can either change it by using the 'Range of Display' scale, or increase it by changing the factor that is on its right : this factor multiplies the frequency range. The smallest possible step between 2 analyzed and displayed frequencies is 2 Hz.

You can choose between 2 types of view :  - a 3D view that shows frequency, intensity and time; It is possible to rotate or translate the display along the X, Y and Z axis, either by keyboard commands, or by mouse (+/- keyboard) commands, or by touch gestures (if enabled). All the analyzed values are kept in memory, wich implies that even if zoom or perspective changes, any value can be retrieved at any time, even when playing is paused.
					   - a 2D 'flat' view : that shows time versus frequency; the intensity however is represented by the amount of red colour;

The perespective can be reset, set to a 'front' view (frequency versus intensity), or set to values preset by the user. The text can be displayed or not, as well as the line scale. A 'pointer' can be displayed, showing a location on the scale and its exact numerical value as well as the intensity of the frequency (fot the 3D view).

You'll find below a summary of the keyboard, mouse and gestures commands. You can also find a summary of this in the menu, at the 'Help'->'Shortcuts' and 'Gestures Shortcuts' items. From the 'Edit'->'Preferences' menu, you can specify : the width of the displaying window (700 to 1200 pixels), the path of the font beeing used, the coordinates of the preset position of the graph, the color of display ('rainbow' option make the color to change so that the differents 'layers' can be distinguished), wether you want Spectrum3d to run in realtime mode (and if so, choose beween SCHED_RR or SCHED_FIFO policy, and set priority). All those preferences are kept in a 'preferences' file that is created at first use of Spectrum3d and placed in the home directory : ~/.spectrum3d/spectrum3d.pref. Everytime Spectrum3d starts, it check some values of this 'preferences' file for consistency; if out of range value is found, a new 'preferences' file with default values is created.

KEYBOARD AND MOUSE SHORTCUTS :
""""""""""""""""""""""""""""""	
Some commands are possible by keyboard or keyborad and mouse combination. Every mouse commands are done with the left click.

Command			Keyboard		Mouse (+/-keyboard)
================================================================================================================================
Play/pause		Space bar
Stop			Escape
Rotate around X axis	Up/dowm			Mouse up/down
Rotate around Y axis	Right/left		Mouse right/left
Rotate around Z axis	'c' + Up/down		'c' + mouse up/down
Translate along X axis	'x' + left/right	'x' + mouse right/left
Translate along Y axis	'y' + up/down		'y' + mouse up/down
Translate along Z axis	'z' + up/down		'z' + mouse up/down
Increase/decrease Gain	'g' + up/down		'g' + mouse up/down
Change starting 
of the zoom		's' + up/down
Move pointer up/down	SHIFT + right/left
Move pointer up/down
    fast		SHIFT + CTRL + right/left
Read small 50 msec 
samples	of a paused
audio file		'v' + left

GESTURES COMMANDS :
"""""""""""""""""""	
Some commands are possible by keyboard or keyborad and mouse combination. Every mouse commands are done with the left click.

Command				Gesture		
=======================================================================
Play/pause			Double Tap
Rotate around X axis		1 finger Drag Up/dowm	
Rotate around Y axis		1 finger Right/left
Rotate around Z axis		2 fingers Rotate
Translate along X axis		2 fingers Right/left
Translate along Y axis		2 fingers Drag Up/dowm
Translate along Z axis		2 fingers Pinch

PREFERENCES :
"""""""""""""


*********************
|  IMPORTANT NOTES  |
*********************

1. Although it is 3D with OpenGL, the drivers for 3D acceleration are not needed (no need for proprietary drivers here!), but it can help to have a better rendering. 

2. Spectrum 3D can use a lot of ressources on your machine, given the real-time analysis and 3D display of the data's. 3 aspects are importants : the power of the processor, the performance of the graphic card (with or without the 3D drivers) and the type of kernel that is beeing used (a preempt or real-time will allow better performance). The amount of needed ressources is proportionnal to the width of the window. If your machine has difficulty, you can use a smaller window size (in Menu -> Edit -> Preferences), wich decreases the needed ressources.

3. RealTime mode will allow Spectrum3d to have priority on other running processes, in order to not be slowed down by them; it won't decrease the amount of CPU and GPU ressources needed to perform the task. IMPORTANT WARNING : If you want to take advantage of this you should edit or create an '/etc/security/limits.d/audio.conf' file and put this inside : 

	@audio   -  rtprio     95
	
then you should add yourself as a member of the 'audio' group :


ChangeLog

2011-10-13  Bernard Victor Delvaux  <nadaeck@hotmail.com>
    Version : Spectrum3d-1.1.0 :
    * display.c : in the 'flat view', the graph is drawed using the GL_QUAD_STRIP instead of a lot of GL_QUADS; this saves a lot of CPU; 
    * display.c : the 'flat view' can be now be viewed at any width if the window, instead of of only 1200 pixels;
    * display.c : a new type of view is introduced : the 3D flat view, that is a kind of compromise between the 3D view and the strict 2D 'flat' view;
    * preferences.c : preferences file 'spretrum3d.pref' had been replaced by a 'spectrum3drc' file; preferences values are now listed in arrays at the top of the prefernences.c file and can be used, checked and handled in a more automatic way;
    * menu.c : clicking OK after just changes the values but doesn't open the preferences file anymore; on the other way, every preferences are written in the spectrum3drc file when the Spectrum3d is closed;
    * equalizer.c : fix a bug with band-pass/band-reject filter (filter couldn't be changed without be set off then on in again);
    * mainwindow.c : 'name' property of the 'jackaudiosrc' element set to "spectrum3d" instead of the default name ("jackaudiosrc0");

2011-08-25  Bernard Victor Delvaux  <nadaeck@hotmail.com>

    Version : Spectrum3d-1.0.0 : 
    * main.c : rewriten to make UI more clear and organized;
    * main.c : zoom is much more precise : range is from 40 Hz to 20000 Hz; first displayed frequency can be any value;
    * display : added a 'pointer' that can be set precisely; the corresponding frequency is then showed;
    * display : added another type of view : the 'flat view' : 2D view in which the intensity of the frequencies are in differents levels of red;
    * display : in the 3D view, frequencies are now drawed as 'LINES_STRIPS' and not 'QUADS' anymore;
    * equalizer.c : new file : set up an equalizer and band-pass/band-reject filter;
    * scale.c (drawPointer and drawPointerText): draw the pointer and its corresponding numerical values;

2011-07-25  Bernard Victor Delvaux  <nadaeck@hotmail.com>

    Version : Spectrum3d-0.2.3 : 
    * geis.c : filtering of number of touches is now effective (Bug#769859 in libutouch-Geis);
    * configure.ac and makefile.am : use of Autoconf and Autoconf for compilation and installation;

2011-05-04  Bernard Victor Delvaux  <nadaeck@hotmail.com>

    Version : Spectrum3d-0.2.2' : 
    * geis.c : Spectrum3d is now multitouch (using the uTouch-Geis API); 
    * menu.c : new file with all the menu and submenu code;
    * testrtpool : new file : allows real time processing of audio;
    * configure and makefile : use of 'configure' and 'makefile' to compile and install.


2011-04-08  Bernard Victor Delvaux  <nadaeck@hotmail.com>

    This is the first release of 'Spectrum3d-0.2.1' : it is now 3D with OpenGL : the image can be rotated or translated along the three axis, even if the diplay is paused.


2011-01-05  Bernard Victor Delvaux  <nadaeck@hotmail.com>

    This is the first release of the software called 'Spectrum-0.1.1', showing only a 'flat view' perspective in a SDL window; no 3D here yet. It support microphone , audio file or jack input.



src/menu.c
/* This file is part of Spectrum3D.

    Spectrum3D is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Spectrum3D is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Spectrum3D.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <gtk/gtk.h>

#include "config.h"

#include "menu.h"

/* Select font */
void selectFont(GtkWidget *comboColor, gpointer data)
{
	GtkFileFilter *filter = gtk_file_filter_new();
	GtkWidget *pFileSelection;
	GtkWidget *pParent;
	pParent = GTK_WIDGET(data);
	gchar *selectedFont;

	pFileSelection = gtk_file_chooser_dialog_new("Select File",
	GTK_WINDOW(pParent),
	GTK_FILE_CHOOSER_ACTION_OPEN,
	GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
	GTK_STOCK_OPEN, GTK_RESPONSE_OK,
	NULL);
	gtk_file_filter_set_name(filter, ".ttf Files");
	gtk_file_filter_add_pattern(filter, "*.ttf");
	gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(pFileSelection), filter);
	gtk_window_set_modal(GTK_WINDOW(pFileSelection), TRUE);

	switch(gtk_dialog_run(GTK_DIALOG(pFileSelection))) {
		case GTK_RESPONSE_OK:
			selectedFont = gtk_file_chooser_get_filename(GTK_FILE_CHOOSER(pFileSelection));
			sprintf(fontPreference, "%s", selectedFont);
			printf("Selected font is %s\n", fontPreference);
		    break;
		default:
		    break;
		}
	gtk_widget_destroy(pFileSelection);
}

/* Preferences window */
void onPreferences(GtkWidget* widget, gpointer data)
{
	GtkWidget *pPreferences, *pNotebook, *pTabLabel;
    	GtkWidget *pSpinWidth, *pLabel, *pCheckRealtime, *pComboPolicy, *pSpinPriority, *pCheckPreset, *pCheckEnableTouch, *spinInterval, *comboColor;
	GtkWidget *pFrame, *pButtonSelectFont;
	gboolean bState;
	int i = 0;
	gchar *sTabLabel[3];
	GtkWidget *pVBox[3];
	for (i = 0; i < 3; i++) {
		pVBox[i] = gtk_vbox_new(TRUE, 0);
		}
	GtkWidget *pHBox[25];
	for (i = 0; i < 25; i++) {
		pHBox[i] = gtk_hbox_new(TRUE, 0);
		}
		
        pPreferences = gtk_dialog_new_with_buttons("Preferences",
        GTK_WINDOW(mainWindow),
        GTK_DIALOG_MODAL,
        GTK_STOCK_OK,GTK_RESPONSE_OK,
        GTK_STOCK_CANCEL,GTK_RESPONSE_CANCEL,
        NULL);
	gtk_window_set_default_size(GTK_WINDOW(pPreferences), 300, 300);

	pNotebook = gtk_notebook_new();
	gtk_box_pack_start(GTK_BOX(GTK_DIALOG(pPreferences)->vbox), pNotebook, TRUE, TRUE, 0);
	gtk_notebook_set_tab_pos(GTK_NOTEBOOK(pNotebook), GTK_POS_TOP);
        gtk_notebook_set_scrollable(GTK_NOTEBOOK(pNotebook), TRUE);

///////////////////////// 1st Notebook : "Display" //////////////////////
	
        sTabLabel[0] = g_strdup_printf("Display");
	pTabLabel = gtk_label_new(sTabLabel[0]);
	gtk_notebook_append_page(GTK_NOTEBOOK(pNotebook), pVBox[0], pTabLabel);
	g_free(sTabLabel[0]);

	/* SpinButton for choosing the width of the main window */
	pFrame = gtk_frame_new("Width of the window");
	pSpinWidth = gtk_spin_button_new_with_range(700, 1500, 50);
	gtk_box_pack_start(GTK_BOX(pVBox[0]), pHBox[0], TRUE, TRUE, 0);	
	gtk_box_pack_start(GTK_BOX(pHBox[0]), pFrame, TRUE, TRUE, 0);
	gtk_container_add(GTK_CONTAINER(pFrame), pSpinWidth);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON(pSpinWidth), (float)presetWidth); 
	
	gtk_box_pack_start(GTK_BOX(GTK_DIALOG(pPreferences)->vbox), pHBox[1], TRUE, FALSE, 0);

	/* Button to select the font */
	pButtonSelectFont = gtk_button_new_with_mnemonic("Select Font"); 
	gtk_box_pack_start(GTK_BOX(pHBox[2]), pButtonSelectFont, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[0]), pHBox[2], FALSE, FALSE, 0);
	g_signal_connect(G_OBJECT(pButtonSelectFont), "clicked", G_CALLBACK(selectFont), NULL);

	gtk_box_pack_start(GTK_BOX(GTK_DIALOG(pPreferences)->vbox), pHBox[3], TRUE, FALSE, 0);

	/* Combo box to change the color of the display */
	pFrame = gtk_frame_new("Color of Display");
	comboColor = gtk_combo_box_new_text();
	gtk_container_add(GTK_CONTAINER(pFrame), comboColor);
	gtk_box_pack_start(GTK_BOX(pHBox[4]), pFrame, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[0]), pHBox[4], FALSE, FALSE, 0);
	gtk_combo_box_append_text(GTK_COMBO_BOX(comboColor), "PURPLE");
	gtk_combo_box_append_text(GTK_COMBO_BOX(comboColor), "RAINBOW");
	gtk_combo_box_set_active(GTK_COMBO_BOX(comboColor), 0);

	/* Check button to save the current position as 'Preset' */
	pCheckPreset = gtk_check_button_new_with_label("Save current position as preset");
	gtk_box_pack_start(GTK_BOX(pHBox[7]), pCheckPreset, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[0]), pHBox[7], FALSE, FALSE, 0);

///////////////////////// 2d Notebook : "Audio" //////////////////////
	
	sTabLabel[1] = g_strdup_printf("Audio");
	pTabLabel = gtk_label_new(sTabLabel[1]);
	gtk_notebook_append_page(GTK_NOTEBOOK(pNotebook), pVBox[1], pTabLabel);
	g_free(sTabLabel[1]);

	/* SpinButton for choosing the speed */
	pFrame = gtk_frame_new("Interval (msec)");
	spinInterval = gtk_spin_button_new_with_range(100, 250, 50);
	 gtk_spin_button_set_value (GTK_SPIN_BUTTON(spinInterval), (gdouble)interval);
	gtk_container_add(GTK_CONTAINER(pFrame), spinInterval);
	gtk_box_pack_start(GTK_BOX(pHBox[11]), pFrame, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[1]), pHBox[11], FALSE, FALSE, 0);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON(spinInterval), interval);

	gtk_box_pack_start(GTK_BOX(pVBox[1]), pHBox[13], FALSE, FALSE, 0);
	//gtk_box_pack_start(GTK_BOX(pVBox[1]), pHBox[14], FALSE, FALSE, 0);

	/* Realtime */
	pCheckRealtime = gtk_check_button_new_with_label("Enable realtime \n(without Jack)");
	gtk_box_pack_start(GTK_BOX(pHBox[15]), pCheckRealtime, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[1]), pHBox[15], TRUE, FALSE, 0);
	//realtime = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(pCheckRealtime));
	if (realtime) {
		gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(pCheckRealtime), TRUE);
		}

	pFrame = gtk_frame_new("Realtime Policy");
	pComboPolicy = gtk_combo_box_new_text();
	gtk_container_add(GTK_CONTAINER(pFrame), pComboPolicy);
	gtk_box_pack_start(GTK_BOX(pHBox[15]), pFrame, TRUE, TRUE, 0);
	gtk_combo_box_append_text(GTK_COMBO_BOX(pComboPolicy), "SCHED_RR");
	gtk_combo_box_append_text(GTK_COMBO_BOX(pComboPolicy), "SCHED_FIFO");
	if (strstr(policyName, "FIFO")){
		gtk_combo_box_set_active(GTK_COMBO_BOX(pComboPolicy), 1);
		}
  	else {
		gtk_combo_box_set_active(GTK_COMBO_BOX(pComboPolicy), 0);
		}
	
	pFrame = gtk_frame_new("Priority");
	pSpinPriority = gtk_spin_button_new_with_range(50, 80, 1);
	gtk_container_add(GTK_CONTAINER(pFrame), pSpinPriority);
	gtk_box_pack_start(GTK_BOX(pHBox[15]), pFrame, FALSE, FALSE, 0);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON(pSpinPriority), priority);

#ifdef HAVE_LIBUTOUCH_GEIS
///////////////////////// 3d Notebook : "Touch" //////////////////////
	
	sTabLabel[2] = g_strdup_printf("Touch");
	pTabLabel = gtk_label_new(sTabLabel[2]);
	gtk_notebook_append_page(GTK_NOTEBOOK(pNotebook), pVBox[2], pTabLabel);
	g_free(sTabLabel[2]);

	pCheckEnableTouch = gtk_check_button_new_with_label("Enable Touch");
	gtk_box_pack_start(GTK_BOX(pHBox[21]), pCheckEnableTouch, TRUE, TRUE, 0);
	gtk_box_pack_start(GTK_BOX(pVBox[2]), pHBox[21], TRUE, FALSE, 0);
	//realtime = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(pCheckRealtime));
	if (enableTouch) {
		gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(pCheckEnableTouch), TRUE);
		}
#endif

//////////////////////// End of the Notebooks //////////////////////////

	gtk_widget_show_all(GTK_DIALOG(pPreferences)->vbox);
	
    	switch (gtk_dialog_run(GTK_DIALOG(pPreferences)))
		{
		case GTK_RESPONSE_OK:
	    		pref = fopen(prefPath, "w+");
			if (pref != NULL) {
				presetWidth = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(pSpinWidth));
					change_adjust(NULL, NULL);
					fprintf(pref, "%d\n", presetWidth);
					width = presetWidth;
					fprintf(pref, "%s\n", fontPreference);
				gint index = gtk_combo_box_get_active(GTK_COMBO_BOX(comboColor));
				colorType = index; // ColorType enum is detailed in menu.h
				bState = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(pCheckPreset));
					if (bState) {
						presetX = X;
							fprintf(pref, "%f\n", X);
						presetY = Y;
							fprintf(pref, "%f\n", Y);
						presetZ = Z;
							fprintf(pref, "%f\n", Z);
						presetAngleH = AngleH;
							fprintf(pref, "%f\n", AngleH);
						presetAngleV = AngleV;
							fprintf(pref, "%f\n", AngleV);
						presetAngleZ = AngleZ;
							fprintf(pref, "%f\n", AngleZ);
					}
					else {
						fprintf(pref, "%f\n", presetX);
						fprintf(pref, "%f\n", presetY);
						fprintf(pref, "%f\n", presetZ);
						fprintf(pref, "%f\n", presetAngleH);
						fprintf(pref, "%f\n", presetAngleV);
						fprintf(pref, "%f\n", presetAngleZ);
						}
				interval = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(spinInterval));
				fprintf(pref, "%d\n", interval);
				realtime = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(pCheckRealtime));
					fprintf(pref, "%d\n", realtime);
				gchar *string = gtk_combo_box_get_active_text(GTK_COMBO_BOX(pComboPolicy));
				sprintf(policyName, "%s", string);
					fprintf(pref, "%s\n", policyName);
				priority = gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(pSpinPriority));
					fprintf(pref, "%d\n", priority);
				enableTouch = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(pCheckEnableTouch));
					fprintf(pref, "%d\n", enableTouch);
				fclose(pref);
			    }
			else {
				printf("*** WARNING ***  Cannot open preferences file\n");
				}
		break;
		case GTK_RESPONSE_CANCEL:
		case GTK_RESPONSE_NONE:
		default:
		    break;
	}
	gtk_widget_destroy(pPreferences);	
}

/* Keyboard and Mouse shortcuts menu */
void onShortcuts (GtkWidget* widget, gpointer data)
{
	GtkWidget *shortcutsWindow, *pListView, *pScrollbar;
	GtkListStore *pListStore;
	GtkTreeViewColumn *pColumn;
	GtkCellRenderer *pCellRenderer;
	gchar *action[15] = {"Play/Pause", "Stop", "Rotate around X axis", "Rotate around Y axis", "Rotate around Z axis", "Translate along X axis", "Translate along Y axis", "Translate along Z axis", "Increase/decrease Gain", "Change starting value of the zoom", "Read small 50 msec samples	of a paused audio file", "Move pointer up/down", "Move pointer up/down fast"	 };
	gchar *keyboard[15] = {"Space bar", "Escape", "Up/Down", "Right/Left", "'c' + Up/Down", "'x' + Left/Right", "'y' + Up/Down", "'z' + Up/Down", "'g' + Up/Down", "'s' + Up/Down", "'v' + Left", "SHIFT + right/left", "SHIFT + CTRL + right/left" };
	gchar *mouse[15] = {"", "", "Mouse Up/Down", "Mouse Right/Left", "'c' + mouse Up/Down", "'x' + mouse Right/Left", "'y' + mouse Up/Down", "'z' + mouse Up/Down", "'g' + mouse Up/Down" };
	gint i;

    	shortcutsWindow = gtk_dialog_new_with_buttons("Keyboard and mouse shortcuts",
        GTK_WINDOW(mainWindow),
        GTK_DIALOG_MODAL,
        GTK_STOCK_OK,GTK_RESPONSE_OK,
        NULL);
	gtk_window_set_default_size(GTK_WINDOW(shortcutsWindow), 700, 400);
	
	pListStore = gtk_list_store_new(N_COLUMN, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING);
	
	for(i = 0 ; i < 15 ; ++i){
		GtkTreeIter pIter;
		gtk_list_store_append(pListStore, &pIter);
		gtk_list_store_set(pListStore, &pIter,
		    ACTION_COLUMN, action[i],
		    KEYBOARD_COLUMN, keyboard[i],
		    MOUSE_COLUMN, mouse[i],
		    -1);
	}
	
	pListView = gtk_tree_view_new_with_model(GTK_TREE_MODEL(pListStore));

	/* 1st column */
        pCellRenderer = gtk_cell_renderer_text_new();
    	pColumn = gtk_tree_view_column_new_with_attributes("ACTION",
        	pCellRenderer,
		"text", ACTION_COLUMN,
		NULL);

    	gtk_tree_view_append_column(GTK_TREE_VIEW(pListView), pColumn);

	/* 2d column */
	pCellRenderer = gtk_cell_renderer_text_new();
   	pColumn = gtk_tree_view_column_new_with_attributes("KEYBORD SHORTCUT",
		pCellRenderer,
		"text", KEYBOARD_COLUMN,
		NULL);

        gtk_tree_view_append_column(GTK_TREE_VIEW(pListView), pColumn);

	/* 3d column */
	pCellRenderer = gtk_cell_renderer_text_new();
   	pColumn = gtk_tree_view_column_new_with_attributes("MOUSE SHORTCUT",
		pCellRenderer,
		"text", MOUSE_COLUMN,
		NULL);

        gtk_tree_view_append_column(GTK_TREE_VIEW(pListView), pColumn);

	pScrollbar = gtk_scrolled_window_new(NULL, NULL);
	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(pScrollbar),
		GTK_POLICY_AUTOMATIC,
		GTK_POLICY_AUTOMATIC);
	gtk_container_add(GTK_CONTAINER(pScrollbar), pListView);

	gtk_container_add(GTK_CONTAINER(GTK_DIALOG(shortcutsWindow)->vbox), pScrollbar);


	gtk_widget_show_all(GTK_DIALOG(shortcutsWindow)->vbox);
	
    	switch (gtk_dialog_run(GTK_DIALOG(shortcutsWindow)))
		{
		case GTK_RESPONSE_OK:
	    		break;
		default:
		    break;
	}
	gtk_widget_destroy(shortcutsWindow);
}
 
/* Gestures menu */
void onGesturesShortcuts (GtkWidget* widget, gpointer data)
{
	GtkWidget *shortcutsWindow, *pListView, *pScrollbar;
	GtkListStore *pListStore;
	GtkTreeViewColumn *pColumn;
	GtkCellRenderer *pCellRenderer;
	gchar *action[8] = {"Play/Pause", "Rotate around X axis                   ", "Rotate around Y axis", "Rotate around Z axis", "Translate along X axis", "Translate along Y axis", "Translate along Z axis"};
	gchar *keyboard[8] = {"Double Tap", "1 Finger Drag Up/Down", "1 Finger Drag Right/Left", "2 Fingers Rotate", "2 Fingers Drag Right/Left", "2 Fingers Drag Up/Down", "2 Fingers Pinch/Spread"};
	gint i;

    	shortcutsWindow = gtk_dialog_new_with_buttons("Keyboard and mouse shortcuts",
        GTK_WINDOW(mainWindow),
        GTK_DIALOG_MODAL,
        GTK_STOCK_OK,GTK_RESPONSE_OK,
        NULL);
	gtk_window_set_default_size(GTK_WINDOW(shortcutsWindow), 500, 300);
	
	pListStore = gtk_list_store_new(N_COLUMN, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING);
	
	for(i = 0 ; i < 8 ; ++i){
		GtkTreeIter pIter;
		gtk_list_store_append(pListStore, &pIter);
		gtk_list_store_set(pListStore, &pIter,
		    ACTION_COLUMN, action[i],
		    KEYBOARD_COLUMN, keyboard[i],
		    -1);
	}
	
	pListView = gtk_tree_view_new_with_model(GTK_TREE_MODEL(pListStore));

	/* 1st column */
        pCellRenderer = gtk_cell_renderer_text_new();
    	pColumn = gtk_tree_view_column_new_with_attributes("ACTION",
        	pCellRenderer,
		"text", ACTION_COLUMN,
		NULL);

    	gtk_tree_view_append_column(GTK_TREE_VIEW(pListView), pColumn);

	/* 2d column */
	pCellRenderer = gtk_cell_renderer_text_new();
   	pColumn = gtk_tree_view_column_new_with_attributes("GESTURE",
		pCellRenderer,
		"text", KEYBOARD_COLUMN,
		NULL);

        gtk_tree_view_append_column(GTK_TREE_VIEW(pListView), pColumn);

	pScrollbar = gtk_scrolled_window_new(NULL, NULL);
	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(pScrollbar),
		GTK_POLICY_AUTOMATIC,
		GTK_POLICY_AUTOMATIC);
	gtk_container_add(GTK_CONTAINER(pScrollbar), pListView);

	gtk_container_add(GTK_CONTAINER(GTK_DIALOG(shortcutsWindow)->vbox), pScrollbar);


	gtk_widget_show_all(GTK_DIALOG(shortcutsWindow)->vbox);
	
    	switch (gtk_dialog_run(GTK_DIALOG(shortcutsWindow)))
		{
		case GTK_RESPONSE_OK:
	    		break;
		default:
		    break;
	}
	gtk_widget_destroy(shortcutsWindow);
}

void onQuickStart(GtkWidget* widget, gpointer data)
{
	GtkWidget *quickStartWindow, *label;
	quickStartWindow = gtk_dialog_new_with_buttons("Spectrum3d : Quick Start",
        GTK_WINDOW(mainWindow),
        GTK_DIALOG_MODAL,
        GTK_STOCK_OK,GTK_RESPONSE_OK,
        NULL);
	gtk_window_set_default_size(GTK_WINDOW(quickStartWindow), 500, 300);

	gchar *quickStartText = " * 2 views are possible : 3D or 2D (flat view);\n * position of display and perspective can be translated or rotated along the 3 axis (see 'Shortcuts' in menu)\n * zoom can be adjusted :\n       - first displayed frequency can have almost any value;\n       - the range of displayed frequencies can vary from 40 to 20000 Hz;\n * scale (text and line) and a pointer pointing to an exact location can be displayed;\n * perspective and zoom can be changed even when playing is paused;";
	label = gtk_label_new (quickStartText);
	gtk_container_add(GTK_CONTAINER(GTK_DIALOG(quickStartWindow)->vbox), label);
	gtk_widget_show_all(GTK_DIALOG(quickStartWindow)->vbox);
	
    	switch (gtk_dialog_run(GTK_DIALOG(quickStartWindow)))
		{
		case GTK_RESPONSE_OK:
	    		break;
		default:
		    break;
	}
	gtk_widget_destroy(quickStartWindow);
}

/* About menu */
void onAbout(GtkWidget* widget, gpointer data)
{
    GtkWidget *pAbout;
    pAbout = gtk_message_dialog_new (GTK_WINDOW(data),
        GTK_DIALOG_MODAL,
        GTK_MESSAGE_INFO,
        GTK_BUTTONS_OK,
        PACKAGE_STRING
        "\nhttp://www.presences.org/spectrum3d");
    gtk_dialog_run(GTK_DIALOG(pAbout));
    gtk_widget_destroy(pAbout);
}



